c/tskit/trees.c

```diff
@@ -1053,9 +1055,10 @@ update_running_sum(tsk_id_t u, double sign, const double *restrict branch_length
 }
 
 static int
-tsk_treeseq_branch_general_stat(tsk_treeseq_t *self, size_t state_dim,
-    double *sample_weights, size_t result_dim, general_stat_func_t *f, void *f_params,
-    size_t num_windows, double *windows, double *result, tsk_flags_t TSK_UNUSED(options))
+tsk_treeseq_branch_general_stat(const tsk_treeseq_t *self, size_t state_dim,
+    const double *sample_weights, size_t result_dim, general_stat_func_t *f,
+    void *f_params, size_t num_windows, const double *windows, double *result,
```

Comment:
I think ``f_params`` can be marked const too? The intention would certainly be for these to be read-only values that are passed through to the summary function.

Reply:
`f_params` can be const if `params` in `general_stat_func_t` is const:
https://github.com/tskit-dev/tskit/blob/5ac0bb430c59d38c32158c4dc798ca544ee20f3f/c/tskit/trees.h#L292

So I think the question comes down to which should be easier:
1) let callers pass immutable f_param or
2) let implementors of a general_stat_func_t function modify the general purpose pass-through param

For 2) I could imagine maybe an implementor wants to keep some parallel running statistic or something. This f_param/param is untyped and it's left to the implementor to do with it whatever they like. Unfortunately I have little sense whether that "whever they like" would always be read-it-only vs sometimes store-something.

Reply:
I'm going to update and play around with those general_stat_func_t params. They could use better names and I'll see if it's easy and sensible to const some of them.

Reply:
Yeah, that's not obvious I guess. The intention was that we're just passing in read-only values and I can't see a good reason why someone would want to update stuff within the  summary function. But then, maybe that's being overly prescriptive.

Is there a distinction between saying *we* don't update this input pointer vs *you* can't?

Reply:
OK, I think I found a good reason to leave param not const:
https://github.com/tskit-dev/tskit/blob/d8a76acbf4529ae897fa731ddeee282cb32c6692/c/tests/test_stats.c#L306
It's test code, but still, it's an example of a summary function wanting to change values that f_param points to.

> Is there a distinction between saying _we_ don't update this input pointer vs _you_ can't?

Not with normal `const` usage, at least that I'm aware of. It's a distinction that makes sense, but I don't think the C language/compiler has something for this. The code could perform some const casting gymnastics and this distinction could be documented. But that doesn't seem worth it.

I've renamed the params to `general_stat_func_t` to readable names I found in `trees.c` and I'm making the `X`, to be renamed `state` parameter to be const, the only one. I'm thinking `f_param` should be left non-const. Custom code is simpler to read and it's more flexible. It's a little wonky. But in a way if something is `void *` it's kind of like saying the param is not typed and those all the fancy powerful C compiler typing is being skipped, including const.